ledger/internal/cow_creatables.go

```diff
@@ -0,0 +1,216 @@
+	}
+	m := make(map[basics.AppIndex]basics.AppParams, len(acct.AppParams))
+	for k, v := range acct.AppParams {
+		m[k] = v
+	}
+	m[aidx] = params
+	acct.AppParams = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) PutAppLocalState(addr basics.Address, aidx basics.AppIndex, state basics.AppLocalState) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AppIndex]basics.AppLocalState, len(acct.AppLocalStates))
+	for k, v := range acct.AppLocalStates {
+		m[k] = v
+	}
+	m[aidx] = state
+	acct.AppLocalStates = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) PutAssetHolding(addr basics.Address, aidx basics.AssetIndex, data basics.AssetHolding) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AssetIndex]basics.AssetHolding, len(acct.Assets))
+	for k, v := range acct.Assets {
+		m[k] = v
+	}
+	m[aidx] = data
+	acct.Assets = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) PutAssetParams(addr basics.Address, aidx basics.AssetIndex, data basics.AssetParams) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AssetIndex]basics.AssetParams, len(acct.AssetParams))
+	for k, v := range acct.AssetParams {
+		m[k] = v
+	}
+	m[aidx] = data
+	acct.AssetParams = m
+	return cs.putAccount(addr, acct)
+}
+
+func (cs *roundCowState) DeleteAppParams(addr basics.Address, aidx basics.AppIndex) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AppIndex]basics.AppParams, len(acct.AppParams))
+	for k, v := range acct.AppParams {
+		m[k] = v
+	}
+	delete(m, aidx)
+	acct.AppParams = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) DeleteAppLocalState(addr basics.Address, aidx basics.AppIndex) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AppIndex]basics.AppLocalState, len(acct.AppLocalStates))
+	for k, v := range acct.AppLocalStates {
+		m[k] = v
+	}
+	delete(m, aidx)
+	acct.AppLocalStates = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) DeleteAssetHolding(addr basics.Address, aidx basics.AssetIndex) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AssetIndex]basics.AssetHolding, len(acct.Assets))
+	for k, v := range acct.Assets {
+		m[k] = v
+	}
+	delete(m, aidx)
+	acct.Assets = m
+	return cs.putAccount(addr, acct)
+}
+func (cs *roundCowState) DeleteAssetParams(addr basics.Address, aidx basics.AssetIndex) error {
+	acct, err := cs.lookup(addr)
+	if err != nil {
+		return err
+	}
+	m := make(map[basics.AssetIndex]basics.AssetParams, len(acct.AssetParams))
+	for k, v := range acct.AssetParams {
+		m[k] = v
+	}
+	delete(m, aidx)
+	acct.AssetParams = m
+	return cs.putAccount(addr, acct)
+}
+
+func (cs *roundCowState) CheckAppLocalState(addr basics.Address, aidx basics.AppIndex) (ok bool, err error) {
```

Comment:
Does this exist because we're planning ahead for a db layer that makes checking existence faster than returning the value?  Or to avoid the function call overhead of returning the value?

Reply:
Empty local state is can exist (empty schema + empty KV) and this is a valid, so the `ok` flag would be needed anyway. Plus, existence check does not perform deserialization so it is a bit faster.

Eventually it will be a two steps check: 1) modified data, 2) a base cow/ledger caches and `COUNT` query.

Reply:
Besides SQL count and copying values, there are also potential performance benefits in different KV stores that would make checking existence faster than fetching & returning a copy of the value, but the basic thing I was trying to capture with the extra CheckX methods was what's the absolute minimum needed from the account DB when evaluating. Maybe HasX is a better name though..

Reply:
I'm totally on board with the purpose, I was mostly checking to make sure that's why it existed.

I do like either HasXXX() or XXXExists() though, as CheckXXX() sounds like it's doing a sanity check or logic check on the values.